src/Lazy/Lazy/Lazy.cs

```diff
@@ -0,0 +1,39 @@
+using System;
+
+namespace Lazy;
+
+/// <summary>
+/// Evaluate supplier once, then return previously calculated result.
+/// </summary>
+/// <typeparam name="T"> Type of supplier result. </typeparam>
+public class Lazy<T> : ILazy<T>
+{
+    private Func<T> _supplier;
+    private bool _isComputed;
+    private T _result;
+
+    /// <summary>
+    /// Create instance of <see cref="Lazy{T}"/>.
+    /// </summary>
+    /// <param name="supplier"> Function to evaluate. </param>
+    /// <exception cref="ArgumentNullException"> Throws exception, when supplier is null. </exception>
+    public Lazy(Func<T> supplier)
+    {
+        _supplier = supplier ?? throw new ArgumentNullException();
```

Comment:
```suggestion
        _supplier = supplier ?? throw new ArgumentNullException(nameof(supplier);
```
А ещё лучше ArgumentNullException.ThrowIfNull